dashboard/modules/job/job_agent.py

```diff
@@ -0,0 +1,380 @@
+            import_path=repr(job_package_dir),
+            redis_address=repr(ip + ":" + str(port)),
+            redis_password=repr(self._redis_password),
+            driver_entry=self._job_info.driver_entry(),
+            driver_args=driver_args)
+        with open(driver_entry_file, "w") as fp:
+            fp.write(driver_code)
+        return driver_entry_file
+
+    async def run(self):
+        log_dir = self._job_info.log_dir()
+        job_id = self._job_info.job_id()
+        python = self._get_current_python()
+        driver_file = self._gen_driver_code()
+        driver_cmd = [python, "-u", driver_file]
+        stdout_file, stderr_file = self._new_log_files(log_dir,
+                                                       f"driver-{job_id}")
+        return await self._start_driver(driver_cmd, stdout_file, stderr_file,
+                                        self._job_info.env())
+
+
+class JobAgent(dashboard_utils.DashboardAgentModule,
+               job_agent_pb2_grpc.JobAgentServiceServicer):
+    def __init__(self, dashboard_agent):
+        super().__init__(dashboard_agent)
+        self._job_table = {}
+
+    async def InitializeJobEnv(self, request, context):
+        job_id = binary_to_hex(request.job_data.job_id)
+        if job_id in self._job_table:
+            logger.info("[%s] Job environment is ready, skip initialization.",
+                        job_id)
+            return job_agent_pb2.InitializeJobEnvReply(
+                status=agent_manager_pb2.AGENT_RPC_STATUS_OK)
+
+        try:
+            job_info = JobInfo(job_id, json.loads(
+                request.job_data.job_payload), self._dashboard_agent.temp_dir,
+                               self._dashboard_agent.log_dir)
+        except json.JSONDecodeError as ex:
+            error_message = str(ex)
+            error_message += f", job_payload:\n{request.job_data.job_payload}"
+            logger.error("[%s] Initialize job environment failed, %s.", job_id,
+                         error_message)
+            return job_agent_pb2.InitializeJobEnvReply(
+                status=agent_manager_pb2.AGENT_RPC_STATUS_FAILED,
+                error_message=error_message)
+        except Exception as ex:
+            logger.exception(ex)
+            return job_agent_pb2.InitializeJobEnvReply(
+                status=agent_manager_pb2.AGENT_RPC_STATUS_FAILED,
+                error_message=traceback.format_exc())
+
+        async def _initialize_job_env():
+            os.makedirs(
+                job_consts.JOB_DIR.format(
+                    temp_dir=job_info.temp_dir(), job_id=job_id),
+                exist_ok=True)
+            await DownloadPackage(job_info,
+                                  self._dashboard_agent.http_session).run()
+            if request.start_driver:
+                logger.info("[%s] Starting driver.", job_id)
+                language = job_info.language()
+                if language == "PYTHON":
```

Comment:
Add a language enum?

Reply:
The enum language is defined in the protobuf. I will change the string to the enum name.

Reply:
Fixed.